src/openloco/input.cpp

```diff
@@ -77,4 +83,786 @@ namespace openloco::input
+                    if (w != nullptr)
+                    {
+                        if (w->is_enabled(_pressedWidgetIndex))
+                        {
+                            auto pressedWidget = &w->widgets[_pressedWidgetIndex];
+
+                            audio::play_sound(audio::sound_id::sound_2, w->x + ((pressedWidget->left + pressedWidget->right) / 2));
+                            w->call_on_mouse_up(_pressedWidgetIndex);
+                        }
+                    }
+                }
+
+                w->call_on_move(_dragLastX, _dragLastY);
+            }
+            break;
+
+            default:
+                break;
+        }
+    }
+
+    // 0x004C8048
+    static void state_normal(mouse_button state, int16_t x, int16_t y, ui::window* window, ui::widget_t* widget, ui::widget_index widgetIndex)
+    {
+        switch (state)
+        {
+            case mouse_button::left_pressed:
+                state_normal_left(x, y, window, widget, widgetIndex);
+                break;
+            case mouse_button::right_pressed:
+                state_normal_right(x, y, window, widget, widgetIndex);
+                break;
+            case mouse_button::released:
+                state_normal_hover(x, y, window, widget, widgetIndex);
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    static void input_widget_over_flatbutton_invalidate()
+    {
+        ui::window_type windowType = _hoverWindowType;
+        uint16_t widgetIdx = _hoverWidgetIdx;
+        uint16_t windowNumber = _hoverWindowNumber;
+
+        if (windowType == ui::window_type::undefined)
+        {
+            ui::windowmgr::invalidate_widget(windowType, windowNumber, widgetIdx);
+            return;
+        }
+
+        ui::window* oldWindow = ui::windowmgr::find(windowType, windowNumber);
+
+        if (oldWindow != nullptr)
+        {
+            oldWindow->call_prepare_draw();
+
+            ui::widget_t* oldWidget = &oldWindow->widgets[widgetIdx];
+            if (
+                oldWidget->type == ui::widget_type::wt_16 || oldWidget->type == ui::widget_type::wt_10 || oldWidget->type == ui::widget_type::wt_9)
+            {
+
+                ui::windowmgr::invalidate_widget(windowType, windowNumber, widgetIdx);
+            }
+        }
+    }
+
+    // 0x004C8098
+    static void state_normal_hover(int16_t x, int16_t y, ui::window* window, ui::widget_t* widget, ui::widget_index widgetIndex)
+    {
+        if (window->type != *_hoverWindowType || window->number != *_hoverWindowNumber || widgetIndex != *_hoverWidgetIdx)
+        {
+            input_widget_over_flatbutton_invalidate();
+            _hoverWindowType = window->type;
+            _hoverWindowNumber = window->number;
+            _hoverWidgetIdx = widgetIndex;
+            input_widget_over_flatbutton_invalidate();
+        }
+
+#if LOCOMOTION
+        if (window != nullptr && widgetIndex != -1)
```

Comment:
`window != nullptr` does not make sense here, you've already dereferenced it in line 498